webchat/utils/chatClient.js

```diff
@@ -0,0 +1,361 @@
+       * @param {string} eventName - name of the event
+       * @param {object} event - event data
+       */
+      emit(eventName, event) {
+        if (this.eventHandlers[eventName]) {
+          this.eventHandlers[eventName].forEach((callback) => {
+            callback(event);
+          });
+        }
+      }
+
+      /**
+       * Mimics node's on/emit event handlers.
+       * Adds a callback to an event.
+       * @param {string} eventName - name of the event
+       * @param {Function} callback - function to be called when event is raised.
+       */
+      on(eventName, callback) {
+        if (!this.eventHandlers[eventName]) {
+          this.eventHandlers[eventName] = [];
+        }
+        this.eventHandlers[eventName].push(callback);
+      }
+
+      /**
+       * Sends a message (JSON string) to another user via the websocket
+       * @param {string} type - type of message
+       * @param {string} to - id of the user to send a message to
+       * @param {object} data - contents of the message
+       */
+      sendToWebsocket(type, to, data) {
+        const msg = { type, from: this.id, to, data }
+        this.ws.send(JSON.stringify(msg));
+      }
+
+      /**
+       * Sends a message (JSON string) to the server, withou a target user.
+       * @param {string} type - type of message
+       * @param {object} data - contents of the message. defaults to empty object;
+       */
+      sendServerMessage(type, data = {}) {
+        this.sendToWebsocket(type, '', data);
+      }
+
+      /**
+       * Create a new Chat object associated to an id of another user
+       * @param {string} id - firebase id of the other user
+       * @param {MediaStream} stream - a Video/Audio stream for the connection
+       *
+       * @returns {Chat} a chat object of this connection
+       * @fires ChatClient#chat-created
+       */
+      createChat(id, stream) {
+        const chat = new Chat(stream, this.id);
+        this.chats[id] = chat;
+
+        /**
+         * Event when a new chat is created.
+         *
+         * @event ChatClient#chat-created
+         * @property {string} id - id of the other user associated with this chat
+         * @property {Chat} chat - Chat object
+         */
+        this.emit('chat-created', { id, chat });
+        return chat;
+      }
+
+      /**
+       * Handles events sent by the websocket server,
+       * parsing the string content on {e.data} as a javascript regular object.
+       * @param {MessageEvent} e - websocket event
+       *
+       * @listens websocket#call-request
+       * @listens websocket#call-answer
+       * @listens websocket#rpc-offer
+       * @listens websocket#rpc-answer
+       * @listens websocket#ice-candidate-received
+       * @listens websocket#user-list-update
+       */
+      handleWebSocket(event) {
+        const msg = JSON.parse(event.data);
+        console.log('ws event: ', msg.type);
+        console.log(msg);
+
+        switch (msg.type) {
```

Comment:
you can create a mapper
mapper = {type: function()}
so you just have to call mapper[type]();